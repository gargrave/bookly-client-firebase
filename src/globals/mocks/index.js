import { authorMocks } from './data/authorMocks';
import { bookBucketMocks } from './data/bookBucketMocks';
import { bookMocks } from './data/bookMocks';
import { profileMocks } from './data/profileMocks';
import { userMocks } from './data/userMocks';

export {
  authorMocks,
  bookBucketMocks,
  bookMocks,
  profileMocks,
  userMocks,
};
