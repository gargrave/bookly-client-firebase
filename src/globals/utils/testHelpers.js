import React from 'react';
import { shallow } from 'enzyme';

export class ComponentBuilder {
  constructor(component, defaultProps) {
    this.component = component;
    this.defaultProps = defaultProps;
  }

  shallowGetComponent(extraProps = {}) {
    const props = {
      ...this.defaultProps,
      ...extraProps,
    };
    return shallow(<this.component {...props} />);
  }
}
