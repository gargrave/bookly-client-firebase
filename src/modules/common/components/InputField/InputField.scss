@import '../../../../styles/_colors';

.error {
  color: $errorColor;
}

input.invalid {
  border-color: $errorColor;
}
