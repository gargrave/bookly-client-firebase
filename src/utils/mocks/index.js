import authorMocks from './data/authors';
import userMocks from './data/users';

export {
  authorMocks,
  userMocks,
};
