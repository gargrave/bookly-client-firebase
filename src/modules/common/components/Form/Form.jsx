// @flow
import React from 'react'
import { any, array, func, bool, string } from 'prop-types'

import { buildClass } from '../../../../utils/cssHelpers'

import Alert from '../Alert/Alert'
import Button from '../Button'
import ButtonRow from '../ButtonRow/ButtonRow'

import styles from './Form.css'

type Props = {
  cancelBtnText?: string,
  children?: any,
  classes?: string[],
  disabled?: boolean,
  onCancel?: Function,
  onSubmit: Function,
  submitBtnText?: string,
  submitDisabled?: boolean,
  topLevelError?: string,
}

const Form = ({
  cancelBtnText,
  children,
  classes = [],
  disabled,
  onCancel,
  onSubmit,
  submitBtnText,
  submitDisabled = false,
  topLevelError,
}: Props) => {
  return (
    <div>
      {topLevelError && <Alert message={topLevelError} type="danger" />}
      <form
        className={buildClass(styles.form, ...(classes || []))}
        onSubmit={onSubmit}
        noValidate
      >
        {children}

        <ButtonRow>
          <Button
            canSubmit={true}
            disabled={submitDisabled || disabled || false}
            onClick={onSubmit}
            position="left"
            text={submitBtnText || 'Submit'}
            type="success"
          />

          {onCancel && (
            <Button
              onClick={onCancel}
              position="right"
              text={cancelBtnText || 'Cancel'}
              type="light"
            />
          )}
        </ButtonRow>
      </form>
    </div>
  )
}

Form.propTypes = {
  cancelBtnText: string,
  children: any,
  classes: array,
  disabled: bool,
  onCancel: func,
  onSubmit: func.isRequired,
  submitBtnText: string,
  submitDisabled: bool,
  topLevelError: string,
}

export default Form
