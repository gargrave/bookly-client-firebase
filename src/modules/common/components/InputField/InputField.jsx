// @flow
import React from 'react'
import { bool, func, number, oneOf, string } from 'prop-types'
import styled from 'react-emotion'

import { colors } from '../../../../styles'
import { clamp } from '../../../../utils/mathHelpers'

type Props = {
  boundValue: string,
  disabled?: boolean,
  error?: string,
  label?: string,
  maxLength?: number,
  name: string,
  onInputChange: Function,
  placeholder?: string,
  type?: string,
}

const acceptableTypes = ['email', 'password', 'search', 'text']

const StyledInput = styled('input')`
  border-color: ${props => props.invalid && colors.error} !important;
`

const StyledError = styled('p')`
  color: ${colors.error};
`

const InputField = ({
  boundValue,
  disabled,
  error,
  label,
  maxLength = 255,
  name,
  onInputChange,
  placeholder,
  type,
}: Props) => {
  return (
    <div className="input-field">
      {label && <label htmlFor={name}>{label}:</label>}

      <StyledInput
        disabled={disabled || false}
        id={name}
        invalid={!!error}
        maxLength={clamp(maxLength, 1, 255)}
        name={name}
        onChange={onInputChange}
        placeholder={placeholder || ''}
        type={type || 'text'}
        value={boundValue}
      />

      {error && <StyledError className="error">{error}</StyledError>}
    </div>
  )
}

InputField.propTypes = {
  boundValue: string.isRequired,
  disabled: bool,
  error: string,
  maxLength: number,
  label: string,
  name: string.isRequired,
  onInputChange: func.isRequired,
  placeholder: string,
  type: oneOf(acceptableTypes),
}

export default InputField
