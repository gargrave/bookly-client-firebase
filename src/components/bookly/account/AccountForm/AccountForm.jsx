// @flow
import React from 'react';
import { bool, func, shape, string } from 'prop-types';

import type { Profile, ProfileErrors } from '../../../../globals/flowtypes';

import Form from '../../../common/Form';
import InputField from '../../../common/InputField';

type Props = {
  disabled?: boolean,
  errors: ProfileErrors,
  onCancel: Function,
  onInputChange: Function,
  onSubmit: Function,
  profile: Profile,
  topLevelError?: string,
  submitDisabled?: boolean,
};

const profileShape = shape({
  firstName: string.isRequired,
  lastName: string.isRequired,
});

const AccountForm = ({
  disabled = false,
  errors,
  onCancel,
  onInputChange,
  onSubmit,
  profile,
  submitDisabled,
  topLevelError,
}: Props) => {
  return (
    <Form
      classes={['account-form']}
      disabled={disabled}
      onCancel={onCancel}
      onSubmit={onSubmit}
      submitDisabled={submitDisabled}
      topLevelError={topLevelError}
    >
      <InputField
        boundValue={profile.firstName}
        disabled={disabled}
        error={errors.firstName}
        label="First name"
        name="firstName"
        onInputChange={onInputChange}
      />

      <InputField
        boundValue={profile.lastName}
        disabled={disabled}
        error={errors.lastName}
        label="Last name"
        name="lastName"
        onInputChange={onInputChange}
      />
    </Form>
  );
};

AccountForm.propTypes = {
  profile: profileShape.isRequired,
  disabled: bool,
  errors: profileShape.isRequired,
  onCancel: func.isRequired,
  onInputChange: func.isRequired,
  onSubmit: func.isRequired,
  submitDisabled: bool,
  topLevelError: string,
};

export default AccountForm;
