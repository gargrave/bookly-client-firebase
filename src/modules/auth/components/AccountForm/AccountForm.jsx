// @flow
import React from 'react';
import { bool, func, shape, string } from 'prop-types';

import type { Profile, ProfileErrors } from '../../../../modules/profiles/flowtypes';

import Form from '../../../../modules/common/components/Form/Form';
import InputField from '../../../../modules/common/components/InputField/InputField';

type Props = {
  disabled?: boolean,
  errors: ProfileErrors,
  onCancel: Function,
  onInputChange: Function,
  onSubmit: Function,
  profile: Profile,
  topLevelError?: string,
  submitDisabled?: boolean,
};

const profileShape = shape({
  firstName: string.isRequired,
  lastName: string.isRequired,
});

const AccountForm = ({
  disabled = false,
  errors,
  onCancel,
  onInputChange,
  onSubmit,
  profile,
  submitDisabled,
  topLevelError,
}: Props) => {
  return (
    <Form
      classes={['account-form']}
      disabled={disabled}
      onCancel={onCancel}
      onSubmit={onSubmit}
      submitDisabled={submitDisabled}
      topLevelError={topLevelError}
    >
      <InputField
        boundValue={profile.firstName}
        disabled={disabled}
        error={errors.firstName}
        label="First name"
        maxLength={255}
        name="firstName"
        onInputChange={onInputChange}
      />

      <InputField
        boundValue={profile.lastName}
        disabled={disabled}
        error={errors.lastName}
        label="Last name"
        maxLength={255}
        name="lastName"
        onInputChange={onInputChange}
      />
    </Form>
  );
};

AccountForm.propTypes = {
  profile: profileShape.isRequired,
  disabled: bool,
  errors: profileShape.isRequired,
  onCancel: func.isRequired,
  onInputChange: func.isRequired,
  onSubmit: func.isRequired,
  submitDisabled: bool,
  topLevelError: string,
};

export default AccountForm;
