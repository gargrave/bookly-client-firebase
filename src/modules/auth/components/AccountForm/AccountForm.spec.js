import { ComponentBuilder } from '../../../../globals/utils/testHelpers';

import { profileMocks } from '../../../../globals/mocks';

import Form from '../../../../modules/common/components/Form/Form';
import InputField from '../../../../modules/common/components/InputField/InputField';

import AccountForm from './AccountForm';

const defaultProps = {
  disabled: false,
  errors: {
    firstName: '',
    lastName: '',
  },
  onCancel: jest.fn(),
  onInputChange: jest.fn(),
  onSubmit: jest.fn(),
  profile: { ...profileMocks[0] },
  submitDisabled: false,
  topLevelError: '',
};

const builder = new ComponentBuilder(
  AccountForm,
  defaultProps,
);

describe('AccountForm', () => {
  let component;

  describe('the basics', () => {
    test('matches the snapshot', () => {
      component = builder.shallowGetComponent();
      expect(component).toMatchSnapshot();
    });

    test('renders correctly', () => {
      component = builder.mountComponent();
      expect(component.find('.bookly-account-form')).toHaveLength(1);
      expect(component.find(Form)).toHaveLength(1);
      expect(component.find(InputField)).toHaveLength(2);
    });
  });
});
