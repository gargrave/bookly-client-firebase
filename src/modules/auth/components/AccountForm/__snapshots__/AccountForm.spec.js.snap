// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountForm the basics matches the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AccountForm
    disabled={false}
    errors={
      Object {
        "firstName": "",
        "lastName": "",
      }
    }
    onCancel={[Function]}
    onInputChange={[Function]}
    onSubmit={[Function]}
    profile={
      Object {
        "created": 2017-07-01T01:16:43.881Z,
        "firstName": "Larry",
        "lastName": "Finklestein",
        "owner": "oiw223h3qndlksadffl3ksdf",
        "updated": 2017-07-01T01:16:43.881Z,
      }
    }
    submitDisabled={false}
    topLevelError=""
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <InputField
          boundValue="Larry"
          disabled={false}
          error=""
          label="First name"
          maxLength={255}
          name="firstName"
          onInputChange={[Function]}
        />,
        <InputField
          boundValue="Finklestein"
          disabled={false}
          error=""
          label="Last name"
          maxLength={255}
          name="lastName"
          onInputChange={[Function]}
        />,
      ],
      "disabled": false,
      "onCancel": [Function],
      "onSubmit": [Function],
      "submitDisabled": false,
      "topLevelError": "",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "boundValue": "Larry",
          "disabled": false,
          "error": "",
          "label": "First name",
          "maxLength": 255,
          "name": "firstName",
          "onInputChange": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "boundValue": "Finklestein",
          "disabled": false,
          "error": "",
          "label": "Last name",
          "maxLength": 255,
          "name": "lastName",
          "onInputChange": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <InputField
            boundValue="Larry"
            disabled={false}
            error=""
            label="First name"
            maxLength={255}
            name="firstName"
            onInputChange={[Function]}
          />,
          <InputField
            boundValue="Finklestein"
            disabled={false}
            error=""
            label="Last name"
            maxLength={255}
            name="lastName"
            onInputChange={[Function]}
          />,
        ],
        "disabled": false,
        "onCancel": [Function],
        "onSubmit": [Function],
        "submitDisabled": false,
        "topLevelError": "",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "boundValue": "Larry",
            "disabled": false,
            "error": "",
            "label": "First name",
            "maxLength": 255,
            "name": "firstName",
            "onInputChange": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "boundValue": "Finklestein",
            "disabled": false,
            "error": "",
            "label": "Last name",
            "maxLength": 255,
            "name": "lastName",
            "onInputChange": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
